src/core/file.cc

```diff
@@ -1051,16 +1059,50 @@ static internal::alignments xfs_alignments(const dioattr& da, unsigned block_siz
     if (S_ISBLK(st.st_mode)) {
-        size_t block_size;
-        auto ret = ::ioctl(fd, BLKBSZGET, &block_size);
-        if (ret == -1) {
+        size_t logical_block_size;
+        if (auto ret = ::ioctl(fd, BLKBSZGET, &logical_block_size); ret == -1) {
             return make_exception_future<shared_ptr<file_impl>>(
                     std::system_error(errno, std::system_category(), "ioctl(BLKBSZGET) failed"));
         }
+        size_t physical_block_size;
+        if (auto ret = ::ioctl(fd, BLKPBSZGET, &physical_block_size); ret == -1) {
+            return make_exception_future<shared_ptr<file_impl>>(
+                    std::system_error(errno, std::system_category(), "ioctl(BLKPBSZGET) failed"));
+        }
+
+        // Query DIO memory alignment using statx (kernel 6.1+)
+        // This gives us the actual DMA buffer alignment requirement (typically 4 bytes for NVMe)
+        size_t memory_dma_alignment = physical_block_size; // fallback to physical_block_size
+        struct statx stx;
+        if (syscall(__NR_statx, fd, "", AT_EMPTY_PATH, STATX_DIOALIGN, &stx) == 0) {
+            if (stx.stx_mask & STATX_DIOALIGN) {
+#ifndef SEASTAR_STATX_NEEDS_DIO_FIELDS
+                // Use kernel-reported memory alignment (stx_dio_mem_align)
+                // Field is available in Linux 6.1+ headers
+                memory_dma_alignment = stx.stx_dio_mem_align;
+#else
+                // On older headers, the field doesn't exist in the struct
+                // The kernel may support it, but we can't access it safely
+                // Fall back to physical_block_size
+#endif
+            }
+        }
+        // If statx fails or doesn't support STATX_DIOALIGN, we fall back to physical_block_size
+
+        // For reads: use logical_block_size (no performance penalty for reading 512-byte blocks from 4K sector disks)
+        size_t read_block_size = logical_block_size;
+        // For writes: use physical_block_size to avoid hardware-level read-modify-write
+        // - physical_block_size: smallest unit a physical storage device can write atomically (e.g., 4096 bytes for Advanced Format disks)
+        // - logical_block_size: smallest unit the storage device can address (typically 512 bytes)
+        //
+        // The Linux kernel only enforces logical_block_size alignment for O_DIRECT (see block/fops.c:blkdev_dio_invalid).
+        // Using physical_block_size avoids RMW at the hardware level.
+        size_t write_block_size = physical_block_size;
+
         internal::fs_info fsi;
-        fsi.block_size = block_size;
+        fsi.block_size = write_block_size;
         fsi.nowait_works = blockdev_nowait_works(st.st_dev);
-        fsi.align = internal::alignments { fsi.block_size, fsi.block_size, fsi.block_size, fsi.block_size };
+        fsi.align = internal::alignments { memory_dma_alignment, read_block_size, write_block_size, write_block_size };
```

Comment:
Please, wrap the block sizes detection logic into `blkdev_alignments()` method (similarly to how it's done with `xfs_alignments()` helper)

Reply:
This has been addressed in the latest revision. The block device alignment detection logic has been extracted into a `blkdev_alignments()` helper function following the same pattern as `xfs_alignments()`.

The refactored code:
- Introduces `blkdev_alignments(int fd, dev_t device_id)` at src/core/file.cc:1078
- Calls it from `make_file_impl()` at line 1130: `auto align = blkdev_alignments(fd, st.st_rdev);`
- Follows the same structure as `xfs_alignments()` with proper error handling

The helper encapsulates all ioctl/statx queries and physical_block_size override logic in one place.